### PR TITLE
Bug 2009791: Fix IBM Cloud config usage

### PR DESCRIPTION
**Pull request by cjschaef**
Add support for IBM Cloud to use configuration values for install
resources, such as from install-config.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2009791

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -65,7 +65,7 @@ require (
 	github.com/openshift/client-go v0.0.0-20210409155308-a8e62c60e930
 	github.com/openshift/cloud-credential-operator v0.0.0-20200316201045-d10080b52c9e
 	github.com/openshift/cluster-api-provider-gcp v0.0.1-0.20201203141909-4dc702fd57a5
-	github.com/openshift/cluster-api-provider-ibmcloud v0.0.0-20210702173623-676faba9895d
+	github.com/openshift/cluster-api-provider-ibmcloud v0.0.0-20210925003227-c0f28b165bf7
```
Comment (staebler) [inline go.mod:68]: I don't see anything meaningful being brought in with this bump of the cluster api provider. How does this relate to the BZ?

File: pkg/asset/installconfig/ibmcloud/mock/ibmcloudclient_generated.go [hunk 2]
```diff
@@ -6,40 +6,41 @@ package mock
...
 	resourcecontrollerv2 "github.com/IBM/platform-services-go-sdk/resourcecontrollerv2"
 	resourcemanagerv2 "github.com/IBM/platform-services-go-sdk/resourcemanagerv2"
 	vpcv1 "github.com/IBM/vpc-go-sdk/vpcv1"
 	gomock "github.com/golang/mock/gomock"
 	ibmcloud "github.com/openshift/installer/pkg/asset/installconfig/ibmcloud"
-	reflect "reflect"
 )
 
-// MockAPI is a mock of API interface
+// MockAPI is a mock of API interface.
```
Comment (staebler) [inline pkg/asset/installconfig/ibmcloud/mock/ibmcloudclient_generated.go:20]: Why are there changes to this file? I don't see any changes to the mock module.

File: pkg/asset/installconfig/ibmcloud/validation.go [hunk 3]
```diff
@@ -55,15 +55,15 @@ func validateMachinePool(client API, platform *ibmcloud.Platform, machinePool *i
 	allErrs := field.ErrorList{}
 
 	if machinePool.InstanceType != "" {
-		allErrs = append(allErrs, validateMachinePoolType(client, machinePool.InstanceType, path)...)
+		allErrs = append(allErrs, validateMachinePoolType(client, machinePool.InstanceType, path.Child("type"))...)
```
Comment (staebler) [inline pkg/asset/installconfig/ibmcloud/validation.go:58]: These are gratuitous changes to the BZ.

File: pkg/destroy/ibmcloud/ibmcloud.go [hunk 4]
```diff
@@ -115,6 +114,7 @@ func (o *ClusterUninstaller) destroyCluster() error {
 	}, {
 		{name: "Cloud Object Storage Instances", execute: o.destroyCOSInstances},
 		{name: "DNS Records", execute: o.destroyDNSRecords},
+		{name: "IAM Authorizations", execute: o.destroyIAMAuthorizations},
```
Comment (staebler) [inline pkg/destroy/ibmcloud/ibmcloud.go:117]: This change is not related to this BZ.

File: pkg/types/ibmcloud/validation/machinepool.go [hunk 5]
```diff
@@ -1,14 +1,23 @@
...
 	"github.com/IBM-Cloud/bluemix-go/crn"
 	"github.com/openshift/installer/pkg/types/ibmcloud"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 )
 
 // ValidateMachinePool validates the MachinePool.
-func ValidateMachinePool(mp *ibmcloud.MachinePool, path *field.Path) field.ErrorList {
+func ValidateMachinePool(platform *ibmcloud.Platform, mp *ibmcloud.MachinePool, path *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	for i, zone := range mp.Zones {
```
Comment (staebler) [inline pkg/types/ibmcloud/validation/machinepool.go:15]: This change is only tangentially related to the BZ.

**Review by staebler — COMMENTED** [review]
Please keep the changes in this PR specific to the BZ. The other changes look fine, so feel free to open additional PRs with those changes.

File: pkg/types/ibmcloud/validation/machinepool.go (hunk 5 above)
Comment (cjschaef) [reply pkg/types/ibmcloud/validation/machinepool.go]: Dropped the change

File: pkg/destroy/ibmcloud/ibmcloud.go (hunk 4 above)
Comment (cjschaef) [reply pkg/destroy/ibmcloud/ibmcloud.go]: Dropped the change

File: pkg/asset/installconfig/ibmcloud/mock/ibmcloudclient_generated.go (hunk 2 above)
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/mock/ibmcloudclient_generated.go]: Reverted the update

File: go.mod (hunk 1 above)
Comment (cjschaef) [reply go.mod]: Reverted the update

File: pkg/types/ibmcloud/validation/machinepool.go [hunk 6]
```diff
@@ -7,7 +7,7 @@ import (
 )
 
 // ValidateMachinePool validates the MachinePool.
-func ValidateMachinePool(mp *ibmcloud.MachinePool, path *field.Path) field.ErrorList {
+func ValidateMachinePool(platform *ibmcloud.Platform, mp *ibmcloud.MachinePool, path *field.Path) field.ErrorList {
```
Comment (staebler) [inline pkg/types/ibmcloud/validation/machinepool.go:10]: What are these validation changes doing? I don't see anything other than passing in an additional parameter.

Comment (cjschaef) [reply pkg/types/ibmcloud/validation/machinepool.go:10]: Yep, good point. Removed the changes since they are not being used here.

File: pkg/tfvars/ibmcloud/ibmcloud.go [hunk 7]
```diff
@@ -6,6 +6,8 @@ import (
 	"github.com/openshift/installer/pkg/tfvars/internal/cache"
 	"github.com/openshift/installer/pkg/types"
 	"github.com/pkg/errors"
+
+	ibmcloudprovider "github.com/openshift/cluster-api-provider-ibmcloud/pkg/apis/ibmcloudprovider/v1beta1"
```
Comment (staebler) [inline pkg/tfvars/ibmcloud/ibmcloud.go:10]: nit: The cluster-api-provider import should be in a section before the installer imports.

```suggestion
	"github.com/pkg/errors"
	ibmcloudprovider "github.com/openshift/cluster-api-provider-ibmcloud/pkg/apis/ibmcloudprovider/v1beta1"

	"github.com/openshift/installer/pkg/tfvars/internal/cache"
	"github.com/openshift/installer/pkg/types"
```

**Review by staebler — APPROVED** [review]
/lgtm

**Review by staebler — APPROVED** [review]
/lgtm

File: pkg/tfvars/ibmcloud/ibmcloud.go (hunk 7 above)
Comment (cjschaef) [reply pkg/tfvars/ibmcloud/ibmcloud.go]: I moved the CAPI import, but left the `errors` where it was (I assume that we not intended in the suggestion), given the alphanumeric order

Comment (staebler) [reply pkg/tfvars/ibmcloud/ibmcloud.go]: It would be better for the `errors` import to be in a section above and separate from the installer imports, but that is a very minor nit and well beyond the scope of the BZ.
